keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1655 @@
+   // the claim is allocated.
+   PotentialNodes []string
+
+   // A change of the PotentialNodes field triggers a check
+   // in the driver on which of those nodes the resource
+   // might be made available. It then excludes nodes
+   // by listing those where that is not the case in
+   // UnsuitableNodes.
+   //
+   // Unsuitable nodes will be ignored by the scheduler
+   // when selecting a node for a Pod. All other nodes are
+   // potential candidates, either because no information
+   // is available yet or because allocation might
+   // succeed.
+   //
+   // This can change, so the driver must refresh
+   // this information periodically and/or after
+   // changing resource allocation for some other
+   // ResourceClaim until a node gets selected by
+   // the scheduler.
+   UnsuitableNodes []string
+
+   // This field will get set by the resource driver to
+   // inform the scheduler where it can schedule Pods
+   // using the ResourceClaim.
+   //
+   // A resource driver may already set this before
+   // the resource is allocated. The scheduler will
+   // then check this field in addition to UnsuitableNodes
+   // to filter out nodes where the resource cannot
+   // be allocated.
+   //
+   // For an allocated resource this field defines
+   // where the resource can be used by Pods.
+   //
+   // Setting this field is optional. If nil, the
+   // resource is available everywhere.
+   AvailableOnNodes *corev1.NodeSelector
+
+   // Arbitrary data returned by the driver after a successful allocation.
+   // This data is passed to the driver for all operations involving
+   // the allocated resource. This is opaque for Kubernetes.
+   // Driver documentation may explain to users how to interpret
+   // this data if needed.
+   //
+   // The attributes must be sufficient to deallocate the resource
+   // because the ResourceClass might not be available anymore
+   // when deallocation starts.
+   Attributes map[string]string
+
+    // UsersLimit determines how many entities are allowed to use this resource
+    // at the same time. The default is 1. -1 enables the usage by an unlimited number
+   // of users. Individual containers in a pod are not counted as users, only the Pod
+   // is.
+   UserLimit int
+
+   // ReservedFor indicates which entities are currently allowed to use the resource.
+   // Usually those are Pods, but other users are also possible.
+   //
+   // A scheduler must add a Pod that it is scheduling. This must be done in an
+   // atomic ResourceClaim update because there might be multiple schedulers working
+   // on different Pods that compete for access to the same ResourceClaim.
+   //
+   // kubelet will check this before allowing a Pod to run because a scheduler might
+   // have missed that step, for example because it doesn't support dynamic resource
+   // allocation or the feature was disabled.
+   ReservedFor []metav1.OwnerReference
+}
+
+type ResourceClaimPhase string
```

Comment:
can we avoid the use of phases versus just named conditions?

Reply:
How is that different? Just the name? Would "state" instead of "phase" be better?

Reply:
we have tried to avoid phases

see discussion on phases here: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties

Reply:
I've gone through that, but it isn't clear to me yet how that would apply here.

It seems to be about supporting several different, unrelated status properties. But here a ResourceClaim can always only be in one "phase" or "state". Perhaps it will be possible to remove the ResourceClaimStatus.Phase field entirely and just infer the "phase" from other fields:
- a resource is allocated if (and only if) it has a non-nil allocation result
- reallocation gets triggered by a boolean field